### PR TITLE
Better shrinkv chunking

### PR DESCRIPTION
**Pull request by jcupitt**
shrinkv used to always fetch single lines from the input image. This kept memuse low, but could cause severe overcomputation if earlier pipeline stages performed coordinate transforms.
    
This PR makes it fetch in larger chunks, up to the tile height, or more for very large shrinks. Larger tiles means less upstream overcomputation.

See https://github.com/libvips/libvips/issues/3600

### CODE REVIEW

**Comment by jcupitt** [comment]
Here's a text program to show the issue:

```python
#!/usr/bin/env python

import sys
import pyvips

image = pyvips.Image.new_from_file(sys.argv[1])
image = image.icc_transform("srgb")
image = image.rotate(6)
image = image.shrinkv(2)
image = image.reducev(2)
image.copy_memory()
```

If you get `colour.c` to show the regions it's computing by uncommenting this `printf()`:

https://github.com/libvips/libvips/blob/8d5f33174c956623c1b7eb82c43c828c0511cc22/libvips/colour/colour.c#L237-L242

Then run that program ^^^ with master you see:

```
$ ./lorenz.py ~/pics/babe.jpg | sort
...
vips_colour_gen: icc_transform, left = 0, top = 0, width = 50, height = 17
vips_colour_gen: icc_transform, left = 0, top = 1, width = 50, height = 17
vips_colour_gen: icc_transform, left = 0, top = 2, width = 50, height = 17
...
```

What's happening:

1. `shrinkv` is requesting a single line from its input.
2. This line is being rotated 6 degrees and becomes 17 lines
3. 50x17 pixels are being mapped through the ICC profile
4. `shrinkv` is fetching one line further down 
5. ... and another 50x17 pixels are colour converted, but with a 16-line overlap

So we overcompute the colour transform by a factor of 16. You can count the number of transforms with:

```
$ ./lorenz.py ~/pics/babe.jpg | wc
   7100   99400  563520
```

This PR makes `shrinkv` fetch in larger chunks, up to a limit of tile height. I now see:

```
$ ./lorenz.py ~/pics/babe.jpg
vips_colour_gen: icc_transform, left = 47, top = 0, width = 158, height = 258
vips_colour_gen: icc_transform, left = 0, top = 0, width = 77, height = 271
vips_colour_gen: icc_transform, left = 174, top = 0, width = 158, height = 244
...
```

And the total number of transforms is:

```
$ ./lorenz.py ~/pics/babe.jpg | wc
     45     630    3599
```

**Comment by jcupitt** [comment]
With master on a larger (6k x 4k) jpg I see:

```
$ time ./lorenz.py ~/pics/nina.jpg

real	0m0.475s
user	0m9.913s
sys	0m0.181s
```

And after this PR:

```
$ time ./lorenz.py ~/pics/nina.jpg

real	0m0.305s
user	0m4.295s
sys	0m0.144s
```

**Comment by kleisauke** [comment]
This appears to be beneficial for #3538 as well.

`x.jpg`:
```console
$ curl -LO  https://github.com/libvips/vips-bench/raw/master/sample2.v
$ vips colourspace sample2.v t1.v srgb
$ vips replicate t1.v t2.v 25 15
$ vips extract_area t2.v x.jpg[strip] 0 0 6000 6000
$ rm t{1,2}.v
```

`test.py`:
```python
#!/usr/bin/env python

import sys
import pyvips

image = pyvips.Image.new_from_file(sys.argv[1], access='sequential')
image = image.sharpen(sigma=8).resize(0.1)
image.write_to_file(sys.argv[2])
```

With master branch:
```console
$ time ./test.py x.jpg x2.jpg

real	0m32.578s
user	0m32.416s
sys	0m0.113s
```

This PR:
```console
$ time ./test.py x.jpg x2.jpg

real	0m5.328s
user	0m5.251s
sys	0m0.090s
```

**Comment by lorenz** [comment]
The performance edge (when jumping over the 4x downscale boundary) in my test case still exists, but it is now only a factor of 10 (instead of 450). This is still quite a bit, but it is definitely mangeable. So this PR is a massive improvement of 45x.

**Comment by jcupitt** [comment]
`shrinkh` has similar chunking code, I'll add a commit to improve that as well, it might fix a bit more overcomputation.

**Comment by jcupitt** [comment]
>  but it is now only a factor of 10 (instead of 450)

Thanks for testing!

**Comment by jcupitt** [comment]
I tried making `shrinkh` fetch 32 scanlines at a time. With this revised test prog (which I think matches the behaviour of `resize`):

```python
#!/usr/bin/env python

import sys
import pyvips

image = pyvips.Image.new_from_file(sys.argv[1])
image = image.icc_transform("srgb")
image = image.rotate(6)
image = image.shrinkv(2)
image = image.reducev(2)
image = image.shrinkh(2)
image = image.reduceh(2)
image.copy_memory()
```

And this PR before the shrinkh change, I see:

```
$ time ./lorenz.py ~/pics/babe.jpg | wc
   1752   24528  138059

real	0m0.261s
user	0m0.709s
sys	0m0.037s
```

And with this commit as well:

```
$ time ./lorenz.py ~/pics/babe.jpg | wc
     57     798    4509

real	0m0.095s
user	0m0.142s
sys	0m0.023s
```

Down from 1752 calls to colour transform to 57.

**Comment by jcupitt** [comment]
Oh heh I tried with @lorenz's test image from the linked issue.

Master makes 378683 calls to colour transform, the `shrinkv` change gets it down to 38383, and with both the `shrinkh` and `shrinkv` patch it's 1210.

I swapped the magic 32 scanline value for `vips__fatstrip_height` (default 16), it feels a bit less random. That test image now runs with 2411 calls to colour transform, so another x10 reduction.

**Comment by jcupitt** [comment]
I tried your test prog too Kleis (though with a different image):

Master: 29s
Shrinkv: 5.2s
Shrinkv + shrinkh: 5.0s

So shrinkh might help a little on this too.

**Comment by lorenz** [comment]
With all changes together the performance edge is now at around 1.5x (another 7x improvement compared to only shrinkv fixes) which is perfectly reasonable. Thank you very much for fixing this so quickly!

**Review by kleisauke — APPROVED** [review]
LGTM! I'll let you press the big green merge button when you're ready.

**Comment by jcupitt** [comment]
OK, let's do it! (and try to fix any disasters later)

**Comment by jcupitt** [comment]
(and thank you for reporting @lorenz -- it's great to fix corner cases like this)

**Pull request merged**
